src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -205,6 +205,15 @@ internal void Start()
             gcTimer = t;
         }
 
+        internal void RemoveStuckGrainFromDirectory(ActivationData activationData)
+        {
+            DeactivateActivations(new List<ActivationData>() {activationData}).Ignore();
```

Comment:
you may need to refactor it a little bit, to move the  CounterStatistic.FindOrCreate(StatisticNames.CATALOG_ACTIVATION_SHUTDOWN_VIA_DEACTIVATE_ON_IDLE).Increment(); outside, or pass an argument which counter to increment. And add another counter for stuck activations.